fc/code_generation.py

```diff
@@ -195,42 +193,48 @@ def derivative_name(deriv):
         })
 
     # Create parameter information dicts, and map of parameter variables to their indices
+    # Parameters are inputs that aren't states, and have a constant RHS
     parameter_info = []
     parameter_variables = {}
-    for i, parameter in enumerate(parameters):
-        variable = model.get_variable_by_ontology_term(parameter.rdf_term)
-        parameter_info.append({
-            'index': i,
-            'local_name': parameter.local_name,
-            'var_name': variable_name(variable),
-            'initial_value': model.get_value(variable),
-        })
-        parameter_variables[variable] = i
+    for pvar in protocol_variables:
+        if pvar.input and pvar.model_variable is not None:
+            eq = model.get_definition(pvar.model_variable)
+            if eq is not None and not eq.lhs.is_Derivative and len(eq.rhs.atoms(VariableDummy)) == 0:
+                i = len(parameter_info)
+                parameter_info.append({
+                    'index': i,
+                    'local_name': pvar.local_name,
+                    'var_name': variable_name(pvar.model_variable),
+                    'initial_value': model.get_value(pvar.model_variable),
+                })
+                parameter_variables[pvar.model_variable] = i
 
     # Create output information dicts
     # Each output is associated either with a variable or a list thereof.
     output_info = []
     output_variables = set()
-    for i, output in enumerate(outputs):
-        term = output.rdf_term
-        variables = get_variables_transitively(model, term)
-        output_variables.update(variables)
-        if len(variables) == 0:
-            raise ValueError('No variable annotated as {} found'.format(term))
-        elif len(variables) == 1:
-            length = None  # Not a vector output
-            var_name = variable_name(variables[0])
-        else:
+    for pvar in protocol_variables:
+        if pvar.output and pvar.model_variable is not None:
+            # Single variable output
+            var_name = variable_name(pvar.model_variable)
+            length = None
+            output_variables.add(pvar.model_variable)
+        elif pvar.output_category and pvar.transitive_variables:
             # Vector output
-            if term in vector_orderings:
-                order = vector_orderings[term]
+            # TODO: Can a variable be both?
```

Comment:
Yes, it can. Membrane voltage is the most common example!

Reply:
Unless you mean can an RDF term be both, in which case the answer is no.

Reply:
I think in actions.py I've left the opportunity occasionally for a term to "is" a variable, while it also has transitive variables. Is that something that can go then?